server/src/main/java/org/opensearch/index/store/CompositeDirectory.java

```diff
@@ -0,0 +1,443 @@
+     * @throws IOException in case of I/O error
+     */
+    @Override
+    public void sync(Collection<String> names) throws IOException {
+        logger.trace("sync() called {}", names);
+        writeLock.lock();
+        try {
+            Collection<String> newLocalFiles = new ArrayList<>();
+            Collection<String> remoteFiles = Arrays.asList(getRemoteFiles());
+            for (String name : names) {
+                if (remoteFiles.contains(name) == false) newLocalFiles.add(name);
+            }
+            logger.trace("Synced files : {}", newLocalFiles);
+            localDirectory.sync(newLocalFiles);
+        } finally {
+            writeLock.unlock();
+        }
+    }
+
+    /**
+     * Ensures that directory metadata, such as recent file renames, are moved to stable storage.
+     * @throws IOException in case of I/O error
+     */
+    @Override
+    public void syncMetaData() throws IOException {
+        logger.trace("syncMetaData() called ");
+        writeLock.lock();
+        try {
+            localDirectory.syncMetaData();
+        } finally {
+            writeLock.unlock();
+        }
+    }
+
+    /**
+     * Renames {@code source} file to {@code dest} file where {@code dest} must not already exist in
+     * the directory.
+     * @throws IOException in case of I/O error
+     */
+    @Override
+    public void rename(String source, String dest) throws IOException {
+        logger.trace("rename() called {}, {}", source, dest);
+        writeLock.lock();
+        try {
+            localDirectory.rename(source, dest);
+            fileCache.remove(localDirectory.getDirectory().resolve(source));
+            cacheFile(dest);
+        } finally {
+            writeLock.unlock();
+        }
+    }
+
+    /**
+     * Opens a stream for reading an existing file.
+     * Check whether the file is present locally or in remote and return the IndexInput accordingly
+     * @param name the name of an existing file.
+     * @throws IOException in case of I/O error
+     */
+    @Override
+    public IndexInput openInput(String name, IOContext context) throws IOException {
+        logger.trace("openInput() called {}", name);
+        writeLock.lock();
+        try {
+            remoteDirectoryBlobStreamReader.setContext(context);
+            /**
+             * We aren't tracking temporary files (created via createTempOutput) currently in FileCache as these are created and then deleted within a very short span of time
+             * We will be reading them directory from the local directory
+             */
+            if (isTempFile(name)) {
+                return localDirectory.openInput(name, context);
+            }
+            /**
+             * Return directly from the FileCache (via TransferManager) if complete file is present
+             */
+            else if (fileCache.get(localDirectory.getDirectory().resolve(name)) != null) {
```

Comment:
why is this block required? I'd expect all interaction with transferManager to occur through the indexInput on line 299.  if this is the non block based remote fetch case, shouldn't that interact with remoteSegmentStoreDirectory directly?  Will this component be used for such cases?

Reply:
> why is this block required? I'd expect all interaction with transferManager to occur through the indexInput on line 299

This is different from 299. Here we have the entire file present in cache and hence are directly returning from here. In 299 entire file is not there, we will be operating on blocks, and the call to fetch the block will be handled in the fetchBlock function of the OnDemandSnapshotIndexInput

>if this is the non block based remote fetch case, shouldn't that interact with remoteSegmentStoreDirectory directly

The goal is to have all the reads being done from the FileCache itself for better implementation of hot/warm migration and stats related to observability etc.